toolkit/tools/pkg/imagecustomizerlib/customizeboot.go

```diff
@@ -95,74 +104,111 @@ func replaceToken(inputGrubCfgContent string, oldToken string, newToken string)
 	if err != nil {
 		return "", "", err
 	}
-
-	linuxFilePathToken := linuxLine.Tokens[1]
-	start := linuxFilePathToken.Loc.Start.Index
-	end := linuxFilePathToken.Loc.End.Index
-
-	oldKernelPath = inputGrubCfgContent[start:end]
-	outputGrubCfgContent = inputGrubCfgContent[:start] + quotedLinuxPath + inputGrubCfgContent[end:]
-
-	return outputGrubCfgContent, oldKernelPath, nil
+	return outputGrubCfgContent, oldKernelPaths[0], nil
 }
 
 // Overrides the path of the initramfs file of the initrd command within a grub config file.
 func setInitrdPath(inputGrubCfgContent string, initrdPath string) (outputGrubCfgContent string, oldInitrdPath string, err error) {
-	quotedInitrdPath := grub.QuoteString(initrdPath)
-
-	line, err := findSingularGrubCommand(inputGrubCfgContent, "initrd")
+	outputGrubCfgContent, oldInitrdPaths, err := setLinuxOrInitrdPathAll(inputGrubCfgContent, initrdCommand, initrdPath, false /*allowMultiple*/)
 	if err != nil {
 		return "", "", err
 	}
+	return outputGrubCfgContent, oldInitrdPaths[0], nil
+}
 
-	if len(line.Tokens) < 2 {
-		return "", "", fmt.Errorf("grub config 'initrd' command is missing file path arg")
+// Appends kernel command-line args to the linux command within a grub config file.
+// If $kernelopts is present, extraCommandLine is inserted before $kernelopts.
+// If $kernelopts is not present, extraCommandLine is appended at the end.
+func appendKernelCommandLineArgsAll(inputGrubCfgContent string, extraCommandLine string,
+	allowMultiple bool, requireKernelOpts bool) (outputGrubCfgContent string, err error) {
+	lines, err := findLinuxOrInitrdLineAll(inputGrubCfgContent, linuxCommand, allowMultiple)
+	if err != nil {
+		return "", err
 	}
 
-	initrdFilePathToken := line.Tokens[1]
-	start := initrdFilePathToken.Loc.Start.Index
-	end := initrdFilePathToken.Loc.End.Index
+	outputGrubCfgContent = inputGrubCfgContent
+	// loop from last to first so that the captured locations from
+	// findGrubCommandAll are not invalidated as reconstructing
+	// outputGrubCfgContent.
+	for i := len(lines) - 1; i >= 0; i-- {
+		line := lines[i]
 
-	oldInitrdPath = inputGrubCfgContent[start:end]
-	outputGrubCfgContent = inputGrubCfgContent[:start] + quotedInitrdPath + inputGrubCfgContent[end:]
+		// Skip the "linux" command and the kernel binary path arg.
+		argTokens := line.Tokens[2:]
 
-	return outputGrubCfgContent, oldInitrdPath, nil
-}
+		insertAt, err := findCommandLineInsertAt(argTokens, requireKernelOpts)
+		if err != nil {
+			return "", err
+		}
 
-// Appends kernel command-line args to the linux command within a grub config file.
-func appendKernelCommandLineArgs(inputGrubCfgContent string, extraCommandLine string) (outputGrubCfgContent string, err error) {
-	_, insertAt, err := getLinuxCommandLineArgs(inputGrubCfgContent)
-	if err != nil {
-		return "", err
+		// Insert args at the end of the line.
+		outputGrubCfgContent = outputGrubCfgContent[:insertAt] + extraCommandLine + " " + outputGrubCfgContent[insertAt:]
```

Comment:
This should be `+ " " + extraCommandLine + " " +`

You could technically drop the extra space if `requireKernelOpts` is `true`. But it probably isn't worth it.

Reply:
fixed.